### PR TITLE
CAD-4581: Unions

### PR DESCRIPTION
**Pull request by dblarons**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "tcomb-builder",
-  "version": "2.0.0",
+  "version": "2.1.0",
```
Comment (liamcain) [inline package.json]: 2.2.0

File: src/primitives/UnionBuilder.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+   * @param {StructBuilder[]} union - List of builders in the union
+   */
+  setUnion(union) {
+    return new this.constructor(this._state.set('_unionBuilders', Immutable.List(union)));
+  }
+
+  /**
+   * Setting fields is not allowed on a UnionBuilder.
+   */
+  // eslint-disable-next-line class-methods-use-this
```
Comment (liamcain) [inline src/primitives/UnionBuilder.js]: What is this for?

File: src/primitives/__tests__/UnionBuilder.spec.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import UnionBuilder from '../UnionBuilder';
+import StructBuilder from '../StructBuilder';
+import TextBuilder from '../TextBuilder';
+
+describe('UnionBuilder', () => {
+  describe('setUnion and setDispatch', () => {
+    it('should create a valid union type', () => {
+      const world = StructBuilder
```
Comment (liamcain) [inline src/primitives/__tests__/UnionBuilder.spec.js]: international?

File: src/primitives/UnionBuilder.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  /**
+   * Set the dispatch function that conditionally switches between builders.
+   * The function should take a `value` parameter which could be any of the
+   * structs defined in the union. It should inspect the fields of the `value`
+   * struct to decide which struct type to return.
+   *
+   * @param {function} dispatch
+   */
+  setDispatch(dispatch) {
+    return new this.constructor(this._state.set('_unionDispatch', dispatch));
```
Comment (liamcain) [inline src/primitives/UnionBuilder.js]: We might want to just call the key `_dispatch`, in case we want to support other types (eg. Intersection).

**Review by liamcain — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (dblarons) [reply package.json]: thanks

File: src/primitives/UnionBuilder.js (hunk 2 above)
Comment (dblarons) [reply src/primitives/UnionBuilder.js]: The linter expects class methods to call `this` somewhere inside the function body. I'm moving it into the linter rules and removing the check here.

File: src/primitives/UnionBuilder.js (hunk 4 above)
Comment (dblarons) [reply src/primitives/UnionBuilder.js]: Good call

File: src/primitives/__tests__/UnionBuilder.spec.js (hunk 3 above)
Comment (dblarons) [reply src/primitives/__tests__/UnionBuilder.spec.js]: done

**Review by liamcain — DISMISSED** [review]

**Review by liamcain — APPROVED** [review]

**Pull request merged**